crates/compiler/mono/src/ir.rs

```diff
@@ -3057,36 +3064,129 @@ fn specialize_external_help<'a>(
     match specialization_result {
-        Ok((proc, layout)) => {
+        Ok((mut proc, layout)) => {
             let top_level = ProcLayout::from_raw_named(env.arena, name, layout);
 
             if procs.is_module_thunk(name.name()) {
                 debug_assert!(top_level.arguments.is_empty());
             }
 
-            // layouts that are (transitively) used in the type of `mainForHost`.
-            let host_exposed_layouts = top_level
-                .arguments
-                .iter()
-                .copied()
-                .chain([top_level.result]);
+            if !host_exposed_aliases.is_empty() {
+                // layouts that are (transitively) used in the type of `mainForHost`.
+                let mut host_exposed_layouts: Vec<_> = top_level
+                    .arguments
+                    .iter()
+                    .copied()
+                    .chain([top_level.result])
+                    .collect_in(env.arena);
+
+                // it is very likely we see the same types across functions, or in multiple arguments
+                host_exposed_layouts.sort();
+                host_exposed_layouts.dedup();
+
+                // TODO: In the future, we will generate glue procs here
+                for in_layout in host_exposed_layouts {
+                    let layout = layout_cache.interner.get(in_layout);
 
-            // TODO: In the future, we will generate glue procs here
-            for in_layout in host_exposed_layouts {
-                let layout = layout_cache.interner.get(in_layout);
+                    let all_glue_procs = generate_glue_procs(
+                        env.home,
+                        env.ident_ids,
+                        env.arena,
+                        &mut layout_cache.interner,
+                        env.arena.alloc(layout),
+                    );
+
+                    // for now, getters are not processed here
+                    let GlueProcs {
+                        getters: _,
+                        extern_names,
+                    } = all_glue_procs;
+
+                    let mut aliases = BumpMap::default();
+
+                    for (id, mut raw_function_layout) in extern_names {
+                        let symbol = env.unique_symbol();
+                        let lambda_name = LambdaName::no_niche(symbol);
+
+                        if false {
+                            raw_function_layout = match raw_function_layout {
+                                RawFunctionLayout::Function(a, mut lambda_set, _) => {
+                                    lambda_set.ret = in_layout;
+                                    RawFunctionLayout::Function(a, lambda_set, in_layout)
+                                }
+                                RawFunctionLayout::ZeroArgumentThunk(x) => {
+                                    RawFunctionLayout::ZeroArgumentThunk(x)
+                                }
+                            };
+                        }
```

Comment:
is this needed?

Reply:
we'll see. We ran into a problem on a branch, but that did not have the latest type checker/ layout fixes.